### DISCUSSION TITLE
followup=original_access_token_is_no_longer_valid

### DISCUSSION THREAD

**Discussion by israaaa270** [ory/hydra#2372] [Q&A] [answered]
I'm running unit testing of hydra, in case `followup=original_access_token_is_no_longer_valid`

I want to know when access token is no longer valid , is depends on expire time or what.
when I run it , I get:

`Error Trace:	oauth2_auth_code_test.go:312
        	Error:      	Should be false
        	Test:       	TestAuthCodeWithDefaultStrategy/case=perform_authorize_code_flow_with_ID_token_and_refresh_tokens/strategy=jwt/followup=successfully_perform_refresh_token_flow/followup=original_access_token_is_no_longer_valid
        	Messages:   	{"active":true,"scope":"hydra offline openid","client_id":"62272d2f-b072-436e-8ef3-36e10c8f6b6e","sub":"aeneas-rekkas","exp":1614578200,"iat":1614578197,"nbf":1614578197,"aud":[],"iss":"http://127.0.0.1:53629/","token_type":"Bearer","token_use":"access_token","ext":{"foo":"bar"}}
               
 --- FAIL: TestAuthCodeWithDefaultStrategy/case=perform_authorize_code_flow_with_ID_token_and_refresh_tokens/strategy=jwt/followup=successfully_perform_refresh_token_flow/followup=original_access_token_is_no_longer_valid (0.00s)
`

**Comment by Benehiko** [comment, marked as answer]
Hi @israaaa270,

Please look at our documentation (https://www.ory.sh/hydra/docs) for clarification on the expiration of the tokens, for this specific setting - it can be found [here](https://www.ory.sh/hydra/docs/guides/token-expiration#access-token-expiration) :)

**Reply by israaaa270** [reply]
I admire with hydra code and learn it. I need to understand Function RevokeSession in persister_oauth.go file.

 when  this errors happened and why handled it : 

`else if err := sqlcon.HandleError(err); err != nil {

		if errors.Is(err, sqlcon.ErrConcurrentUpdate) {

			return errors.Wrap(fosite.ErrSerializationFailure, err.Error())

		} else if strings.Contains(err.Error(), "Error 1213") { // InnoDB Deadlock?

			return errors.Wrap(fosite.ErrSerializationFailure, err.Error())

		}`
in Function RevokeSession.

`func (p *Persister) revokeSession(ctx context.Context, id string, table tableName) error {

	/* #nosec G201 table is static */

	if err := p.Connection(ctx).RawQuery(

		fmt.Sprintf("DELETE FROM %s WHERE request_id=?", OAuth2RequestSQL{Table: table}.TableName()),

		id,

	).Exec(); errors.Is(err, sql.ErrNoRows) {

		return errorsx.WithStack(fosite.ErrNotFound)

	} else if err := sqlcon.HandleError(err); err != nil {

		if errors.Is(err, sqlcon.ErrConcurrentUpdate) {

			return errors.Wrap(fosite.ErrSerializationFailure, err.Error())

		} else if strings.Contains(err.Error(), "Error 1213") { // InnoDB Deadlock?

			return errors.Wrap(fosite.ErrSerializationFailure, err.Error())

		}

		return err

	}

	return nil

}`

**Reply by tacurran** [reply]
@israaaa270 hello, Can you help us here with the exact path/file names and perhaps state your questions... Lets assume its oath2.go 
https://github.com/ory/hydra/blob/b46a14cd6d260a7dee748de34abfea54908f1a0b/persistence/sql/persister_oauth2.go#L260
persistence/sql/persister_oauth2.go:260

**Reply by israaaa270** [reply]
yes exactly , in hydra/persistence/sql/persister_oauth2.go

**Reply by tacurran** [reply]
So in Go you have many different ways to manage errors, which are treated as values: 
errors.Is(err, sqlcon.ErrConcurrentUpdate) see the github.com/pkg/errors
Function errors.Wrap gives some context to the error found.
For the Fosite package look here: https://pkg.go.dev/github.com/ory/fosite
ErrSerializationFailure is an error indicating that the transactional capable storage could not guarantee consistency of Update & Delete operations on the same rows between multiple sessions. ErrSerializationFailure = errors.New("The request could not be completed due to concurrent access")
If you know all that and your question is related to something else for instance sec or db, please be specific.
